src/cirrus/builtins/functions/process/lambda_function.py

```diff
@@ -1,65 +1,88 @@
-                )
-                for c in _payloads:
-                    c['process'] = dict_merge(
-                        c['process'],
-                        payload['process_update'],
-                    )
-            ProcessPayloads(_payloads).process(replace=True)
-        elif payload.get('type', '') == 'Feature':
-            # If Item, create ProcessPayload and
-            # use default process for that collection
-            if payload['collection'] not in PROCESSES.keys():
-                raise ValueError(
-                    "Default process not provided for "
-                    f"collection {payload['collection']}",
-                )
-            payload_json = {
-                'type': 'FeatureCollection',
-                'features': [payload],
-                'process': PROCESSES[payload['collection']]
-            }
-            payloads.append(ProcessPayload(payload_json, update=True))
-        else:
+    failures = []
+    messages = {}
+    for message in utils.normalize_event(event):
+
+        try:
+            payload = utils.extract_record(message)
+        except Exception as e:
+            logger.exception('Failed to extract record: %s', json.dumps(message))
+            failures.append(message)
+
+        # if the payload has a URL in it then we'll fetch it from S3
+        try:
+            payload = utils.payload_from_s3(payload)
+        except NoUrlError:
+            pass
+
+        logger.debug('payload: %s', defer(json.dumps, payload))
+
+        try:
             payloads.append(ProcessPayload(payload, update=True))
+        except Exception:
+            logger.exception('Failed to convert to ProcessPayload: %s', json.dumps(payload))
+            failures.append(payload)
+
+        if is_sqs_message(message):
+            try:
+                messages[payload['id']].append(message)
+            except KeyError:
+                messages[payload['id']] = [message]
 
+    processed_ids = set()
     if len(payloads) > 0:
-        _payloads = ProcessPayloads(payloads)
-        _payloads.process()
+        processed = ProcessPayloads(payloads).process()
+        logger.error(processed)
+        processed_ids = set(pid for state in processed.keys() for pid in processed[state])
+
+    logger.error(messages)
+    successful_sqs_messages = [
+        message
+        for _id in processed_ids
+        for message in messages.pop(_id, [])
+    ]
+    failures += list(messages.values())
+
+    logger.error(successful_sqs_messages)
+    logger.error(failures)
+
+    if failures:
+        # If we have partial failure, then we want to delete all
+        # successfully processed messages from the queue, so they
+        # won't be reprocessed again. We don't need to do this if
+        # we have no failures, as SQS will delete the messages for
+        # us if we exit successfully.
+        for message in successful_sqs_messages:
+            try:
+                utils.delete_from_queue(message)
```

Comment:
preferable to use [delete_message_batch()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html#SQS.Client.delete_message_batch) here (in batches of 10) rather than one at a time, which can slow this down

Reply:
I considered that initially, but it seemed like premature optimization. That is, chunking the successful messages into blocks of 10 adds complexity, and I don't think we'd often be failing here, so this won't run very often.

That said, you are probably right that we should do it anyway. As I reflect on it more, it seems like we either have only a handful of messages to process at once and they'll all fit into one batch delete call, or we'll have a bunch of messages and need to be using the batch call for performance reasons.

I will put it on the todo list to circle back around on this.

Reply:
I have updated this with a batch method.